# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: media/webrtc/trunk/webrtc/modules/desktop_capture/linux/window_finder_x11.cc
# Commit: 3f0981f37c1c
# Full Hash: 3f0981f37c1c499ad52016e526239b390703ba2a
# Author: Dan Minor <dminor@mozilla.com>
# Date: 2019-04-17 21:47:29
# Regressor Bug: 1496359
# File Overlap Count: 1
# Description:
#   Bug 1496359 - Add PipeWire support to desktop capture; r=ng
#   
#   This is an import of upstream commit 318da51f99f91e3de1192f29d7f1824958f9f13e.
#   
#   Depends on D27367
# ==============================================================================

diff -r ceed4e0b3373 -r 3f0981f37c1c media/webrtc/trunk/webrtc/modules/desktop_capture/linux/window_finder_x11.cc
--- /dev/null	Thu Jan 01 00:00:00 1970 +0000
+++ b/media/webrtc/trunk/webrtc/modules/desktop_capture/linux/window_finder_x11.cc	Mon Apr 15 17:26:27 2019 +0000
@@ -0,0 +1,51 @@
+/*
+ *  Copyright (c) 2017 The WebRTC project authors. All Rights Reserved.
+ *
+ *  Use of this source code is governed by a BSD-style license
+ *  that can be found in the LICENSE file in the root of the source
+ *  tree. An additional intellectual property rights grant can be found
+ *  in the file PATENTS.  All contributing project authors may
+ *  be found in the AUTHORS file in the root of the source tree.
+ */
+
+#include "modules/desktop_capture/window_finder_x11.h"
+
+#include "modules/desktop_capture/x11/window_list_utils.h"
+#include "rtc_base/checks.h"
+#include "rtc_base/ptr_util.h"
+
+namespace webrtc {
+
+WindowFinderX11::WindowFinderX11(XAtomCache* cache)
+    : cache_(cache) {
+  RTC_DCHECK(cache_);
+}
+
+WindowFinderX11::~WindowFinderX11() = default;
+
+WindowId WindowFinderX11::GetWindowUnderPoint(DesktopVector point) {
+  WindowId id = kNullWindowId;
+  GetWindowList(cache_,
+                [&id, this, point](::Window window) {
+                  DesktopRect rect;
+                  if (GetWindowRect(this->cache_->display(), window, &rect) &&
+                      rect.Contains(point)) {
+                    id = window;
+                    return false;
+                  }
+                  return true;
+                });
+  return id;
+}
+
+// static
+std::unique_ptr<WindowFinder> WindowFinder::Create(
+    const WindowFinder::Options& options) {
+  if (options.cache == nullptr) {
+    return nullptr;
+  }
+
+  return rtc::MakeUnique<WindowFinderX11>(options.cache);
+}
+
+}  // namespace webrtc